python/tvm/relax/backend/pattern_registry.py

```diff
@@ -0,0 +1,70 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Pattern registry for BYOC backends"""
+
+from typing import List, Mapping, Optional, Tuple, Union
+
+import tvm
+from tvm.relax.dpl import DFPattern
+from tvm.runtime import Object
+
+from . import _ffi_api
+
+
+@tvm._ffi.register_object("relax.backend.PatternRegistryEntry")
+class PatternRegistryEntry(Object):
+    name: str
+    pattern: DFPattern
+    arg_patterns: Mapping[str, DFPattern]
```

Comment:
what's the use case of `arg_patterns`?

Reply:
It can be used in the backend codegen to identify input arg expressions in a robust way. The codegen can do something like
```python
match_result = extract_matched_expr(...)
attrs.update({f"{arg_name}_idx": find_arg_index(match_result[arg_pattern] 
              for arg_name, arg_pattern in entry.arg_patterns.items()})
```